### PR TITLE
feat: Allow CORS at SiteLevel

### PR DESCRIPTION
**Pull request by faztp12**
What type of a PR is this?
- [x] New Feature Submissions
---

CORS can be a very helpful feature for sites hosted on static hosting providers. I would like to propose a config/flag to allow CORS on specific sites. By specifying `cors_config` either in the common site config / site config enables this.
The approach taken for the headers allows SocketIO connections to establish from a different domain,

All the following 6 CORS Response headers are supported in this PR:
- [Access-Control-Allow-Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin)
- [Access-Control-Expose-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers)
- [Access-Control-Max-Age](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age)
- [Access-Control-Allow-Credentials](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials)
- [Access-Control-Allow-Methods](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods)
- [Access-Control-Allow-Headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers)

An example site_config configuration:
```json
"cors_config": {
  "*": { 
    "enabled": 1,
    "allow_credentials": 1,
    "headers": ["X-Frappe-SiteName"],
    "methods": ["GET", "POST"],
    "max_age": 864000,
    "expose_headers": []
  },
  "http://localhost": {
    "enabled": 1,
    "methods": ["GET"]
  },
  "https://myrandomsite.com": {
    "enabled": 1
  }
}
```

The generated nginx configuration will have `nginx_maps` to resolve the header values for each origin.
```
map $http_origin $cors_headers_twwhxds {
	http://localhost 'X-Frappe-Site-Name, X-Client-Site';
	http://randomsite.com 'X-Frappe-Site, X-Custom-Header';	
}
```

In the server block,
```
# CORS Headers
proxy_hide_header Access-Control-Allow-Origin;
proxy_hide_header Access-Control-Allow-Credentials;
add_header Access-Control-Allow-Origin $cors_origins_twwhxds always;
add_header Access-Control-Allow-Headers $cors_headers_twwhxds always;
add_header Access-Control-Allow-Methods $cors_methods_twwhxds always;
add_header Access-Control-Max-Age $cors_max_age_twwhxds always;
add_header Access-Control-Allow-Credentials $cors_allow_creds_twwhxds always;
```

TODO:
- [x] Site level Origin CORS Configuration
- [x] Support for wildcard origin
- [ ] Tests  
- [ ] Updating Documentations  
Unfortunately I was unable to find docs related to wildcard_ssls or configs related to nginx in this repo, is it in frappe/frappe ?

### CODE REVIEW

**Comment by faztp12** [comment]
@adityahase I have update the PR, along with the description.
Do provide your thoughts on the direction this PR is headed 😄

**Comment by revant** [comment]
For reference:

Alternative to programmatically set CORS headers, it works in dev bench.

https://discuss.erpnext.com/t/cors-on-localhost/60116/7

**Review by adityahase — CHANGES_REQUESTED** [review]
We're maintaing docs at https://github.com/frappe/frappe_docs/blob/master/frappe_docs/www/docs/user/en/bench/index.md.


Please include some tests.

File: bench/config/nginx.py [hunk 1]
```diff
@@ -278,3 +290,98 @@ def get_limit_conn_shared_memory():
...
+			if isinstance(v, list):
+				v = "\"{}\"".format(", ".join(v))
+			elif prop == "allow_credentials":
+				v = "true" if v else "false"
+			parsed_config[prop][origin] = v
+
+		if not parsed_config["max_age"].get(origin):
+			parsed_config["max_age"][origin] = 864000
+
+	return parsed_config
```
[the remark this answers is not in the document]
Comment (faztp12) [reply bench/config/nginx.py:331]: Yep, let me check what can be done 👍

**Comment by faztp12** [comment]
@adityahase should we consider a python approach instead of modifying it in nginx ?
Something similar to what @revant mentioned ?

File: bench/config/nginx.py [hunk 2]
```diff
@@ -278,3 +290,98 @@ def get_limit_conn_shared_memory():
...
+			for k, v in obj1.items():
+				if k == "random_string":
+					continue
+				if not _compare(v, obj2[k]):
+					return False
+			return True
+		elif isinstance(obj1, (list, tuple)):
+			return set(obj1) == set(obj2)
+		else:
+			return obj1 == obj2
```
[the remark this answers is not in the document]
Comment (faztp12) [reply bench/config/nginx.py]: @adityahase please check the updates..

**Comment by adityahase** [comment]
@faztp12 Please add tests.

**Comment by faztp12** [comment]
@adityahase ok 👍

**Comment by revant** [comment]
this got merged into framework.

https://github.com/frappe/frappe/pull/12183

**Pull request closed without merging**